cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/AppMetadataStoreTest.java

```diff
@@ -479,6 +481,140 @@ public void testgetRuns() throws Exception {
+                                   AppFabricTestHelper.createSourceId(sourceId.incrementAndGet()));
+
+        // one run in each stopped state
+        for (ProgramRunStatus runStatus : ProgramRunStatus.values()) {
+          if (!runStatus.isEndState()) {
+            continue;
+          }
+          runId = programId.run(RunIds.generate());
+          store.recordProgramProvisioning(runId, Collections.emptyMap(), Collections.emptyMap(),
+                                          AppFabricTestHelper.createSourceId(sourceId.incrementAndGet()), ARTIFACT_ID);
+          store.recordProgramProvisioned(runId, 3, AppFabricTestHelper.createSourceId(sourceId.incrementAndGet()));
+          twillRunId = UUID.randomUUID().toString();
+          store.recordProgramStart(runId, twillRunId, Collections.emptyMap(),
+                                   AppFabricTestHelper.createSourceId(sourceId.incrementAndGet()));
+          store.recordProgramStop(runId, System.currentTimeMillis(), runStatus, null,
+                                  AppFabricTestHelper.createSourceId(sourceId.incrementAndGet()));
+        }
+      });
+    }
+
+    Set<ProgramRunStatus> activeStates = new HashSet<>();
+    activeStates.add(ProgramRunStatus.PENDING);
+    activeStates.add(ProgramRunStatus.STARTING);
+    activeStates.add(ProgramRunStatus.RUNNING);
+    // check active runs per namespace
+    for (NamespaceId namespace : Arrays.asList(ns1, ns2)) {
+      txnl.execute(() -> {
+        Map<ProgramRunId, RunRecordMeta> activeRuns = store.getActiveRuns(namespace);
+
+        // we expect 3 runs per program, with 4 programs in each namespace
+        Map<ProgramId, Set<ProgramRunStatus>> expected = new HashMap<>();
+        expected.put(namespace.app(app1).mr(program1), activeStates);
+        expected.put(namespace.app(app1).mr(program2), activeStates);
+        expected.put(namespace.app(app2).mr(program1), activeStates);
+        expected.put(namespace.app(app2).mr(program2), activeStates);
+
+        Map<ProgramId, Set<ProgramRunStatus>> actual = new HashMap<>();
+        actual.put(namespace.app(app1).mr(program1), new HashSet<>());
+        actual.put(namespace.app(app1).mr(program2), new HashSet<>());
+        actual.put(namespace.app(app2).mr(program1), new HashSet<>());
+        actual.put(namespace.app(app2).mr(program2), new HashSet<>());
+        for (Map.Entry<ProgramRunId, RunRecordMeta> activeRun : activeRuns.entrySet()) {
+          ProgramId programId = activeRun.getKey().getParent();
+          Assert.assertTrue("Unexpected program returned: " + programId,
+                            actual.containsKey(activeRun.getKey().getParent()));
+          actual.get(programId).add(activeRun.getValue().getStatus());
+        }
+
+        Assert.assertEquals(expected, actual);
+      });
+    }
+
+    // check active runs per app
+    for (ApplicationId app : Arrays.asList(ns1.app(app1), ns1.app(app2), ns2.app(app1), ns2.app(app2))) {
```

Comment:
can get the appId's from transforming the programId's into appId's and collect them as a set

Reply:
is the concern if we add program ids but don't modify this list? I can create an app id collection first then create the program id collection from that.